### ISSUE TITLE
[css-contain-4] c-v: auto elements painted into snapshots for View Transitions

### ISSUE THREAD

**Issue by khushalsagar** [w3c/csswg-drafts#8060]
This was a part of #7874 but wasn't resolved at the last meeting where this issue was discussed. The edit also needs to go into the contain spec so filing another issue.

If an element is participating in a transition, any element in its sub-tree (which has c-v: auto) should become relevant to the user. This makes sense given that the named element will be painted into a snapshot. This snapshot's position and scale is animated during the transition (by the pseudo-element its drawn in). So all its content (including the c-v: auto elements) could be in the viewport.

Given that the snapshot for all non-root elements paints the entire DOM tree underneath, it'll make all c-v: auto elements underneath a non-root named element relevant to the user. It's only the root (for which the snapshot is clipped) which can have c-v: auto elements underneath that are clipped out of the snapshot and hence don't need to be made relevant to the user.

Proposed Resolution: An element with content-visibility: auto is relevant to the user if it is painted into self or an ancestor's snapshot during a view transition.

@vmpstr @frivoal fyi.

**Comment by tabatkins** [comment]
This makes sense to me. 

Would it make sense to explicitly capture the clipped/unclipped snapshot distinction and tie it to that? Or do we expect only the root to ever do this? (We can always change our mind later, this is just about the theoretical organization.)

**Comment by khushalsagar** [comment]
Nested elements can also clip because of hw constraints (like max texture size). The proposed resolution I put was intended to capture the clipped/unclipped distinction without explicitly tie-ing it to the root. If the snapshot doesn't cover the full ink overflow rectangle of an element, then some elements won't be painted. If c-v:auto is not painted because of this clipping then it doesn't need to be made relevant.

**Comment by vmpstr** [comment]
For the proposal there are two parts here:
1. The element is relevant to the user if it contains within its subtree an element that participates in a view transition, e.g.
```html
<div style="content-visibility: auto"> <!-- relevant -->
  <div id=participant style="view-transition-name: foo"></div>
</div>
```

2. The element is relevant to the user if it is inside a subtree of an element that participates in a view transition, e.g.
```html
<div id=participant style="view-transition-name: foo">
  <div style="content-visibility: auto"></div> <!-- relevant -->
</div>
```

Both are needed to have correct visual representation of the transitioning element when drawn via the pseudo elements

**Comment by khushalsagar** [comment]
Case 1 was discussed in #7874 with the resolution that if an element's contents are being skipped then its descendants can not participate in a view transition. The discussion is captured [here](https://github.com/w3c/csswg-drafts/issues/7874#issuecomment-1309141936).

So this issue is limited to case 2. Sounds like we're in agreement that c-v: auto elements in this case will become relevant to the user. @vmpstr are you ok with limiting relevance to "if it is painted into self or an ancestor's snapshot". So an engine could avoid it if the c-v: auto element has been clipped from the snapshot?

**Comment by vmpstr** [comment]
The issue discussed in #7874 was about discovering elements that are being skipped by c-v auto ancestors. Here, the issue is that once an element _does_ participate in the transition, any ancestors that have a `content-visibility: auto` are relevant to the user. So both cases are needed here.

I'm not sure what you mean by the following:
> So an engine could avoid it if the c-v: auto element has been clipped from the snapshot?

Can you clarify

**Comment by khushalsagar** [comment]
> The issue discussed in https://github.com/w3c/csswg-drafts/issues/7874 was about discovering elements that are being skipped by c-v auto ancestors.

Right, so if an element has an ancestor being skipped by c-v auto during the discovery phase it can not participate in a transition. Are you concerned about the case where we decide that an element will participate in the transition. But c-v: auto gets added to an ancestor during the transition animations? It conceptually seems similar to transition constraints being broken mid-transition, like the `display: none` scenario [here](https://github.com/w3c/csswg-drafts/issues/7882#issuecomment-1315767620). So we could do the same thing and skip the transition?

> Can you clarify

The root snapshot is clipped to the snapshot viewport, described in the spec [here](https://drafts.csswg.org/css-view-transitions-1/#ref-for-computing-the-interest-rectangle:~:text=Render%20the%20region%20of%20the%20element%20and%20the%20top%20layer%20that%20intersects%20the%20snapshot%20viewport). Anything that's been clipped out of the snapshot, the browser can continue skipping.

**Comment by khushalsagar** [comment]
Ok. Thanks for clarifying. So a summary and proposed resolution for the 2 cases is:

- An element with c-v: auto has a descendant with a non-none `view-transition-name`. The element did not skip its contents when discovering elements which participate in a transition but does so mid-transition. That would cause the descendant to not paint. So we either force the ancestor to stay relevant throughout the transition or abort it (similar to an ancestor adding `display: none`).
   **Proposed resolution**: An element with c-v: auto is relevant to the user if it contains within its subtree an element that participates in a view transition.

- An element with c-v: auto has an ancestor which is participating in a transition. So the descendant's painting is required for generating the ancestor's image.
   **Proposed resolution**: An element with c-v: auto is relevant to the user if it is inside the subtree of an element that participates in a view transition and is being painted into this ancestor's snapshot.

Lmk if the summary and wording of the proposed resolution sound good.

**Comment by vmpstr** [comment]
> An element with c-v: auto is relevant to the user if it is inside the subtree of an element that participates in a view transition and is being painted into this ancestor's snapshot.

Not sure if the last part `and is being painted into this ancestor's snapshot` is required. What case are you trying to capture here? IOW when is it not being painted into some shared ancestor?

**Comment by khushalsagar** [comment]
> Not sure if the last part and is being painted into this ancestor's snapshot is required. What case are you trying to capture here? IOW when is it not being painted into some shared ancestor?

The root snapshot is clipped to the snapshot viewport, described in the spec [here](https://drafts.csswg.org/css-view-transitions-1/#ref-for-computing-the-interest-rectangle:~:text=Render%20the%20region%20of%20the%20element%20and%20the%20top%20layer%20that%20intersects%20the%20snapshot%20viewport). The c-v: auto element's closest ancestor, which is participating in the view transition, is the root. The root snapshot is clipped to the snapshot viewport. If the c-v: auto element is outside the snapshot viewport (which is likely if its skipping its contents) then we don't need to force it to become relevant to the user. Since it not being rendered into the root snapshot anyway.

**Comment by khushalsagar** [comment]
Posting a summary and updated proposal from an offline discussion:

1. C-V: auto element with a descendant participating in transition
After the resolution at https://github.com/w3c/csswg-drafts/issues/7874#issuecomment-1309141936, if an element skips its contents then its descendants are not eligible to participate in a view transition. The descendants are skipped when discovering participating elements [here](https://www.w3.org/TR/css-view-transitions-1/#setup-view-transition-algorithm) : "If any [flat tree](https://drafts.csswg.org/css-scoping-1/#flat-tree) ancestor of this element [skips its contents](https://drafts.csswg.org/css-contain-2/#skips-its-contents), then [continue](https://infra.spec.whatwg.org/#iteration-continue)."

   So this case is limited to the ancestor of a participating element becoming skippable mid-transition. There is a conceptual intersection between this and the discussion in #7882. If we used the same strategy, we'd skip the transition. But given that skipping a c-v: auto element's contents is dependent on browser heuristics, it's possible for the case to occur without a developer error. For example, more content loading pushing an element down. So the preference is to err on avoiding visual abruptness (by aborting the transition) and keeping the c-v: auto element unlocked throughout the transition.

   **Proposed Resolution**: Element with c-v: auto continues to stay relevant to the user if it has descendants which participate in a transition.

2. C-V: auto element with ancestor participating in transition
This case has a c-v: auto element which is being captured in it's ancestor's image [here](https://www.w3.org/TR/css-view-transitions-1/#capture-the-image). The c-v: auto element itself could be offscreen and be skipping its contents. But it's painting could come onscreen based on the geometry of the pseudo-element which will display the snapshot.

   There is a performance vs visual correctness trade-off here. We'd continue skipping the contents to retain the performance optimization that c-v: auto is supposed to provide. But make the element relevant given the possibility that it's painting *could* be displayed onscreen. We're erring on the side of prioritizing performance since developers are more likely to notice visual correctness issues if they need the element to stay relevant. And they have an easy way to ensure that by switching from c-v: auto to c-v: visible.

   **Proposed Resolution**: Element with c-v: auto is not forced to become relevant to the user if painted in an ancestor's snapshot for view transitions.

We still need to clarify what happens if an elements gets `visibility: hidden` or `content-visibility: hidden` mid transition, will file a separate issue for that.

**Comment by vmpstr** [comment]
I just want to highlight that Khushal's last comment for case 2 has a different resolution from the rest of the discussion on this issue, and is the current proposed resolution

**Comment by astearns** [comment]
RESOLVED: Element with c-v: auto continues to stay relevant to the user if it has descendants which participate in a transition.

RESOLVED: Element with c-v: auto is not forced to become relevant to the user if painted in an ancestor's snapshot for view transitions.
